src/tsd.c

```diff
@@ -25,6 +25,10 @@ bool tsd_booted = false;
 DWORD tsd_tsd;
 tsd_wrapper_t tsd_boot_wrapper = {false, TSD_INITIALIZER};
 bool tsd_booted = false;
+#if JEMALLOC_WIN32_TLSGETVALUE2
+extern TGV2 tlsgetvalue2 = NULL;
```

Comment:
Thanks for the patch @mcfi ! 

You probably don't mean `extern` here? Maybe rename to `tlsgetvalue2ptr` to be more distinguished from TlsGetValue2? Or is this really defined somewhere else that we are only referencing (in which case, will the extern cause trouble if it's not defined anywhere)? 

Also please squash all the commits into a single one.

Reply:
thank you for reviewing my change :) I removed the "extern"s you pointed out. In addition, instead of renaming tlsgetvalue2 to tlsgetvalue2ptr, I chose tls_get_value2 to make the name easier to read and consistent with other variables (no _ptr suffix). Hope it's ok to you.

Could you please advise on how to squash all commits into a single one? Should I close this PR and send out a new one?

Reply:
`git rebase -i` to squash locally and `git push --force` to the remote will do the job.

Reply:
@mcfi `tls_get_value2` sounds good to me. BTW it looks we don't have to keep a global tgv2_mod, instead it can simply be a stack variable?

re: squashing. Like Guangli mentioned, after doing so locally, force push to this branch (same PR) will work.

Reply:
One way to optimize away the NULL pointer function check:

Having a global `tls_get_value_ptr` which is initialized to either `TlsGetValue2` (if available) or fall back to `TlsGetValue_handle_error` which is defined as below:
```
TlsGetValue_handle_error (tsd) {
        DWORD error = GetLastError();
        wrapper = (tsd_wrapper_t *) TlsGetValue(tsd);
        SetLastError(error);
}
```

Which is possible because the two variants of TLS get value share the same function signature. This way it's always an unconditional function call.

Reply:
> @mcfi `tls_get_value2` sounds good to me. BTW it looks we don't have to keep a global tgv2_mod, instead it can simply be a stack variable?
> 
> re: squashing. Like Guangli mentioned, after doing so locally, force push to this branch (same PR) will work.

With tgv_mod being a stack variable, the moment when the initialization function returns, the stack variable is gone, and in case some users need to call FreeLibrary against tgv_mod during application exit, they can't.. Speaking of that, is there any destruction code for jemalloc, if so, I can add a FreeLibrary call there to match LoadLibrary.

> One way to optimize away the NULL pointer function check:
> 
> Having a global `tls_get_value_ptr` which is initialized to either `TlsGetValue2` (if available) or fall back to `TlsGetValue_handle_error` which is defined as below:
> 
> ```
> TlsGetValue_handle_error (tsd) {
>         DWORD error = GetLastError();
>         wrapper = (tsd_wrapper_t *) TlsGetValue(tsd);
>         SetLastError(error);
> }
> ```
> 
> Which is possible because the two variants of TLS get value share the same function signature. This way it's always an unconditional function call.

This may have a bigger perf penalty on older Windows without TlsGetValue2 support if an application is built with some sort of control-flow integrity (e.g., Control-Flow Guard). Right now, TlsGetValue is called with an indirect call through a read-only import table, which doesn't have to be guarded by CFI. However, if we use a function pointer for TlsGetValue, in the worst case we add a CFI check before dereferencing that pointer, which is much more expensive than a simple conditional branch.

Reply:
Thanks for the detailed answers @mcfi. 

> Speaking of that, is there any destruction code for jemalloc, if so, I can add a FreeLibrary call there to match LoadLibrary.

No library destruction support in jemalloc currently.